static/app/utils/useApi.tsx

```diff
@@ -0,0 +1,41 @@
+import {useEffect, useState} from 'react';
+
+import {Client} from 'app/api';
+
+type Options = {
+  /**
+   * Enabling this option will disable clearing in-flight requests when the
+   * component is unmounted.
+   *
+   * This may be useful in situations where your component needs to finish up
+   * some where the client was passed into some type of action creator and the
+   * component is unmounted.
+   */
+  persistInFlight?: boolean;
+  /**
+   * An existing API client may be provided.
+   *
+   * This is a continent way to re-use clients and still inherit the
+   * persistInFlight configuration.
+   */
+  api?: Client;
+};
+
+/**
+ * Returns an API client that will have it's requests canceled when the owning
+ * React component is unmounted (may be disabled via options).
+ */
+function useApi({persistInFlight, api: providedApi}: Options = {}) {
+  const [localApi] = useState(() => new Client());
```

Comment:
The alternative here is to have a useRef and initialize it with `null` and then immediately after check if it needs to be constructed.

Something like this https://github.com/facebook/react/issues/14490#issuecomment-449729465

However, it sounds like using `useState` with the lazy initialization is more ergonomic and has no extra cost

Reply:
can we create it in useEffect? I don't like that a client is always created regardless if one is provided

Reply:
We can't because then we don't guarantee it's created.

But what I can do, is change this to a useRef and lazily instantiate it like I suggested above. Then we can avoid instantiation if we provided one.